File: 876-Middle-of-the-Linked-List.py
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr, count = head, 0
        while curr:
            curr = curr.next
            count += 1

        slow, fast = head, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next

        if count % 2 == 0:
            return slow.next
        return slow

File: neetcode/876-Middle-of-the-Linked-List.ts
function middleNode(head: ListNode | null): ListNode | null {
  let n = 0, m = 0

  let curr = head
  while (curr) {
    curr = curr.next
    n++
  }

  m = Math.floor(n / 2)
  curr = head
  while (m--) {
    curr = curr.next
  }

  return curr
};